## Overview

This notebook will show you how to create and query a table or DataFrame that you uploaded to DBFS. [DBFS](https://docs.databricks.com/user-guide/dbfs-databricks-file-system.html) is a Databricks File System that allows you to store data for querying inside of Databricks. This notebook assumes that you have a file already inside of DBFS that you would like to read from.

This notebook is written in **Python** so the default cell type is Python. However, you can use different languages by using the `%LANGUAGE` syntax. Python, Scala, SQL, and R are all supported.

In [0]:
# File location and type
file_location = "/FileStore/tables/vgsales.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

Rank Name Platform Year Genre Publisher NA_Sales EU_Sales JP_Sales Other_Sales Global_Sales 1 Wii Sports Wii 2006 Sports Nintendo 41.49 29.02 3.77 8.46 82.74 2 Super Mario Bros. NES 1985 Platform Nintendo 29.08 3.58 6.81 0.77 40.24 3 Mario Kart Wii Wii 2008 Racing Nintendo 15.85 12.88 3.79 3.31 35.82 4 Wii Sports Resort Wii 2009 Sports Nintendo 15.75 11.01 3.28 2.96 33 5 Pokemon Red/Pokemon Blue GB 1996 Role-Playing Nintendo 11.27 8.89 10.22 1 31.37 6 Tetris GB 1989 Puzzle Nintendo 23.2 2.26 4.22 0.58 30.26 7 New Super Mario Bros. DS 2006 Platform Nintendo 11.38 9.23 6.5 2.9 30.01 8 Wii Play Wii 2006 Misc Nintendo 14.03 9.2 2.93 2.85 29.02 9 New Super Mario Bros. Wii Wii 2009 Platform Nintendo 14.59 7.06 4.7 2.26 28.62 10 Duck Hunt NES 1984 Shooter Nintendo 26.93 0.63 0.28 0.47 28.31 11 Nintendogs DS 2005 Simulation Nintendo 9.07 11 1.93 2.75 24.76 12 Mario Kart DS DS 2005 Racing Nintendo 9.81 7.57 4.13 1.92 23.42 13 Pokemon Gold/Pokemon Silver GB 1999 Role-Playing Nintendo 9 6.18 7.2 0.71 23.1 14 Wii Fit Wii 2007 Sports Nintendo 8.94 8.03 3.6 2.15 22.72 15 Wii Fit Plus Wii 2009 Sports Nintendo 9.09 8.59 2.53 1.79 22 16 Kinect Adventures! X360 2010 Misc Microsoft Game Studios 14.97 4.94 0.24 1.67 21.82 17 Grand Theft Auto V PS3 2013 Action Take-Two Interactive 7.01 9.27 0.97 4.14 21.4 18 Grand Theft Auto: San Andreas PS2 2004 Action Take-Two Interactive 9.43 0.4 0.41 10.57 20.81 19 Super Mario World SNES 1990 Platform Nintendo 12.78 3.75 3.54 0.55 20.61 20 Brain Age: Train Your Brain in Minutes a Day DS 2005 Misc Nintendo 4.75 9.26 4.16 2.05 20.22 21 Pokemon Diamond/Pokemon Pearl DS 2006 Role-Playing Nintendo 6.42 4.52 6.04 1.37 18.36 22 Super Mario Land GB 1989 Platform Nintendo 10.83 2.71 4.18 0.42 18.14 23 Super Mario Bros. 3 NES 1988 Platform Nintendo 9.54 3.44 3.84 0.46 17.28 24 Grand Theft Auto V X360 2013 Action Take-Two Interactive 9.63 5.31 0.06 1.38 16.38 25 Grand Theft Auto: Vice City PS2 2002 Action Take-Two Interactive 8.41 5.49 0.47 1.78 16.15 26 Pokemon Ruby/Pokemon Sapphire GBA 2002 Role-Playing Nintendo 6.06 3.9 5.38 0.5 15.85 27 Pokemon Black/Pokemon White DS 2010 Role-Playing Nintendo 5.57 3.28 5.65 0.82 15.32 28 Brain Age 2: More Training in Minutes a Day DS 2005 Puzzle Nintendo 3.44 5.36 5.32 1.18 15.3 29 Gran Turismo 3: A-Spec PS2 2001 Racing Sony Computer Entertainment 6.85 5.09 1.87 1.16 14.98 30 Call of Duty: Modern Warfare 3 X360 2011 Shooter Activision 9.03 4.28 0.13 1.32 14.76 31 Pokémon Yellow: Special Pikachu Edition GB 1998 Role-Playing Nintendo 5.89 5.04 3.12 0.59 14.64 32 Call of Duty: Black Ops X360 2010 Shooter Activision 9.67 3.73 0.11 1.13 14.64 33 Pokemon X/Pokemon Y 3DS 2013 Role-Playing Nintendo 5.17 4.05 4.34 0.79 14.35 34 Call of Duty: Black Ops 3 PS4 2015 Shooter Activision 5.77 5.81 0.35 2.31 14.24 35 Call of Duty: Black Ops II PS3 2012 Shooter Activision 4.99 5.88 0.65 2.52 14.03 36 Call of Duty: Black Ops II X360 2012 Shooter Activision 8.25 4.3 0.07 1.12 13.73 37 Call of Duty: Modern Warfare 2 X360 2009 Shooter Activision 8.52 3.63 0.08 1.29 13.51 38 Call of Duty: Modern Warfare 3 PS3 2011 Shooter Activision 5.54 5.82 0.49 1.62 13.46 39 Grand Theft Auto III PS2 2001 Action Take-Two Interactive 6.99 4.51 0.3 1.3 13.1 40 Super Smash Bros. Brawl Wii 2008 Fighting Nintendo 6.75 2.61 2.66 1.02 13.04 41 Call of Duty: Black Ops PS3 2010 Shooter Activision 5.98 4.44 0.48 1.83 12.73 42 Animal Crossing: Wild World DS 2005 Simulation Nintendo 2.55 3.52 5.33 0.88 12.27 43 Mario Kart 7 3DS 2011 Racing Nintendo 4.74 3.91 2.67 0.89 12.21 44 Halo 3 X360 2007 Shooter Microsoft Game Studios 7.97 2.83 0.13 1.21 12.14 45 Grand Theft Auto V PS4 2014 Action Take-Two Interactive 3.8 5.81 0.36 2.02 11.98 46 Pokemon HeartGold/Pokemon SoulSilver DS 2009 Action Nintendo 4.4 2.77 3.96 0.77 11.9 47 Super Mario 64 N64 1996 Platform Nintendo 6.91 2.85 1.91 0.23 11.89 48 Gran Turismo 4 PS2 2004 Racing Sony Computer Entertainment 3.01 0.01 1.1 7.53 11.66 49 Super Mario Galaxy Wii 2007 Platform Nintendo 6.16 3.4 1

In [0]:
import numpy as np
import pandas as pd
from pyspark.sql.types import *
from pyspark.sql import SQLContext
from operator import add
from pyspark.sql.functions import *


In [0]:
# Create a view or table

temp_table_name = "vgsales_csv"

df.createOrReplaceTempView(temp_table_name)

In [0]:
%sql

/* Query the created temp table in a SQL cell */

select * from `vgsales_csv`

Rank Name Platform Year Genre Publisher NA_Sales EU_Sales JP_Sales Other_Sales Global_Sales 1 Wii Sports Wii 2006 Sports Nintendo 41.49 29.02 3.77 8.46 82.74 2 Super Mario Bros. NES 1985 Platform Nintendo 29.08 3.58 6.81 0.77 40.24 3 Mario Kart Wii Wii 2008 Racing Nintendo 15.85 12.88 3.79 3.31 35.82 4 Wii Sports Resort Wii 2009 Sports Nintendo 15.75 11.01 3.28 2.96 33 5 Pokemon Red/Pokemon Blue GB 1996 Role-Playing Nintendo 11.27 8.89 10.22 1 31.37 6 Tetris GB 1989 Puzzle Nintendo 23.2 2.26 4.22 0.58 30.26 7 New Super Mario Bros. DS 2006 Platform Nintendo 11.38 9.23 6.5 2.9 30.01 8 Wii Play Wii 2006 Misc Nintendo 14.03 9.2 2.93 2.85 29.02 9 New Super Mario Bros. Wii Wii 2009 Platform Nintendo 14.59 7.06 4.7 2.26 28.62 10 Duck Hunt NES 1984 Shooter Nintendo 26.93 0.63 0.28 0.47 28.31 11 Nintendogs DS 2005 Simulation Nintendo 9.07 11 1.93 2.75 24.76 12 Mario Kart DS DS 2005 Racing Nintendo 9.81 7.57 4.13 1.92 23.42 13 Pokemon Gold/Pokemon Silver GB 1999 Role-Playing Nintendo 9 6.18 7.2 0.71 23.1 14 Wii Fit Wii 2007 Sports Nintendo 8.94 8.03 3.6 2.15 22.72 15 Wii Fit Plus Wii 2009 Sports Nintendo 9.09 8.59 2.53 1.79 22 16 Kinect Adventures! X360 2010 Misc Microsoft Game Studios 14.97 4.94 0.24 1.67 21.82 17 Grand Theft Auto V PS3 2013 Action Take-Two Interactive 7.01 9.27 0.97 4.14 21.4 18 Grand Theft Auto: San Andreas PS2 2004 Action Take-Two Interactive 9.43 0.4 0.41 10.57 20.81 19 Super Mario World SNES 1990 Platform Nintendo 12.78 3.75 3.54 0.55 20.61 20 Brain Age: Train Your Brain in Minutes a Day DS 2005 Misc Nintendo 4.75 9.26 4.16 2.05 20.22 21 Pokemon Diamond/Pokemon Pearl DS 2006 Role-Playing Nintendo 6.42 4.52 6.04 1.37 18.36 22 Super Mario Land GB 1989 Platform Nintendo 10.83 2.71 4.18 0.42 18.14 23 Super Mario Bros. 3 NES 1988 Platform Nintendo 9.54 3.44 3.84 0.46 17.28 24 Grand Theft Auto V X360 2013 Action Take-Two Interactive 9.63 5.31 0.06 1.38 16.38 25 Grand Theft Auto: Vice City PS2 2002 Action Take-Two Interactive 8.41 5.49 0.47 1.78 16.15 26 Pokemon Ruby/Pokemon Sapphire GBA 2002 Role-Playing Nintendo 6.06 3.9 5.38 0.5 15.85 27 Pokemon Black/Pokemon White DS 2010 Role-Playing Nintendo 5.57 3.28 5.65 0.82 15.32 28 Brain Age 2: More Training in Minutes a Day DS 2005 Puzzle Nintendo 3.44 5.36 5.32 1.18 15.3 29 Gran Turismo 3: A-Spec PS2 2001 Racing Sony Computer Entertainment 6.85 5.09 1.87 1.16 14.98 30 Call of Duty: Modern Warfare 3 X360 2011 Shooter Activision 9.03 4.28 0.13 1.32 14.76 31 Pokémon Yellow: Special Pikachu Edition GB 1998 Role-Playing Nintendo 5.89 5.04 3.12 0.59 14.64 32 Call of Duty: Black Ops X360 2010 Shooter Activision 9.67 3.73 0.11 1.13 14.64 33 Pokemon X/Pokemon Y 3DS 2013 Role-Playing Nintendo 5.17 4.05 4.34 0.79 14.35 34 Call of Duty: Black Ops 3 PS4 2015 Shooter Activision 5.77 5.81 0.35 2.31 14.24 35 Call of Duty: Black Ops II PS3 2012 Shooter Activision 4.99 5.88 0.65 2.52 14.03 36 Call of Duty: Black Ops II X360 2012 Shooter Activision 8.25 4.3 0.07 1.12 13.73 37 Call of Duty: Modern Warfare 2 X360 2009 Shooter Activision 8.52 3.63 0.08 1.29 13.51 38 Call of Duty: Modern Warfare 3 PS3 2011 Shooter Activision 5.54 5.82 0.49 1.62 13.46 39 Grand Theft Auto III PS2 2001 Action Take-Two Interactive 6.99 4.51 0.3 1.3 13.1 40 Super Smash Bros. Brawl Wii 2008 Fighting Nintendo 6.75 2.61 2.66 1.02 13.04 41 Call of Duty: Black Ops PS3 2010 Shooter Activision 5.98 4.44 0.48 1.83 12.73 42 Animal Crossing: Wild World DS 2005 Simulation Nintendo 2.55 3.52 5.33 0.88 12.27 43 Mario Kart 7 3DS 2011 Racing Nintendo 4.74 3.91 2.67 0.89 12.21 44 Halo 3 X360 2007 Shooter Microsoft Game Studios 7.97 2.83 0.13 1.21 12.14 45 Grand Theft Auto V PS4 2014 Action Take-Two Interactive 3.8 5.81 0.36 2.02 11.98 46 Pokemon HeartGold/Pokemon SoulSilver DS 2009 Action Nintendo 4.4 2.77 3.96 0.77 11.9 47 Super Mario 64 N64 1996 Platform Nintendo 6.91 2.85 1.91 0.23 11.89 48 Gran Turismo 4 PS2 2004 Racing Sony Computer Entertainment 3.01 0.01 1.1 7.53 11.66 49 Super Mario Galaxy Wii 2007 Platform Nintendo 6.16 3.4 1

In [0]:
# With this registered as a temp view, it will only be available to this particular notebook. If you'd like other users to be able to query this table, you can also create a table from the DataFrame.
# Once saved, this table will persist across cluster restarts as well as allow various users across different notebooks to query this data.
# To do so, choose your table name and uncomment the bottom line.

permanent_table_name = "vgsales_csv"

# df.write.format("parquet").saveAsTable(permanent_table_name)

## Video Games Sales Analysis

---

The dataset contains information regarding video games sales throughout different regions, together with information about the year released, the console, the genre and the publisher.

The **objective** of the project is to make questions about the dataset and try to answer them using Spark. 

--- 

- **Masters program:** MIP Politecnico di Milano - Business Analytics & Big Data
- **Course**: Cloud Technologies and Big Data Frameworks 
- **Profesor**: Danilo Ardagna

In [0]:
display(df)

Rank Name Platform Year Genre Publisher NA_Sales EU_Sales JP_Sales Other_Sales Global_Sales 1 Wii Sports Wii 2006 Sports Nintendo 41.49 29.02 3.77 8.46 82.74 2 Super Mario Bros. NES 1985 Platform Nintendo 29.08 3.58 6.81 0.77 40.24 3 Mario Kart Wii Wii 2008 Racing Nintendo 15.85 12.88 3.79 3.31 35.82 4 Wii Sports Resort Wii 2009 Sports Nintendo 15.75 11.01 3.28 2.96 33 5 Pokemon Red/Pokemon Blue GB 1996 Role-Playing Nintendo 11.27 8.89 10.22 1 31.37 6 Tetris GB 1989 Puzzle Nintendo 23.2 2.26 4.22 0.58 30.26 7 New Super Mario Bros. DS 2006 Platform Nintendo 11.38 9.23 6.5 2.9 30.01 8 Wii Play Wii 2006 Misc Nintendo 14.03 9.2 2.93 2.85 29.02 9 New Super Mario Bros. Wii Wii 2009 Platform Nintendo 14.59 7.06 4.7 2.26 28.62 10 Duck Hunt NES 1984 Shooter Nintendo 26.93 0.63 0.28 0.47 28.31 11 Nintendogs DS 2005 Simulation Nintendo 9.07 11 1.93 2.75 24.76 12 Mario Kart DS DS 2005 Racing Nintendo 9.81 7.57 4.13 1.92 23.42 13 Pokemon Gold/Pokemon Silver GB 1999 Role-Playing Nintendo 9 6.18 7.2 0.71 23.1 14 Wii Fit Wii 2007 Sports Nintendo 8.94 8.03 3.6 2.15 22.72 15 Wii Fit Plus Wii 2009 Sports Nintendo 9.09 8.59 2.53 1.79 22 16 Kinect Adventures! X360 2010 Misc Microsoft Game Studios 14.97 4.94 0.24 1.67 21.82 17 Grand Theft Auto V PS3 2013 Action Take-Two Interactive 7.01 9.27 0.97 4.14 21.4 18 Grand Theft Auto: San Andreas PS2 2004 Action Take-Two Interactive 9.43 0.4 0.41 10.57 20.81 19 Super Mario World SNES 1990 Platform Nintendo 12.78 3.75 3.54 0.55 20.61 20 Brain Age: Train Your Brain in Minutes a Day DS 2005 Misc Nintendo 4.75 9.26 4.16 2.05 20.22 21 Pokemon Diamond/Pokemon Pearl DS 2006 Role-Playing Nintendo 6.42 4.52 6.04 1.37 18.36 22 Super Mario Land GB 1989 Platform Nintendo 10.83 2.71 4.18 0.42 18.14 23 Super Mario Bros. 3 NES 1988 Platform Nintendo 9.54 3.44 3.84 0.46 17.28 24 Grand Theft Auto V X360 2013 Action Take-Two Interactive 9.63 5.31 0.06 1.38 16.38 25 Grand Theft Auto: Vice City PS2 2002 Action Take-Two Interactive 8.41 5.49 0.47 1.78 16.15 26 Pokemon Ruby/Pokemon Sapphire GBA 2002 Role-Playing Nintendo 6.06 3.9 5.38 0.5 15.85 27 Pokemon Black/Pokemon White DS 2010 Role-Playing Nintendo 5.57 3.28 5.65 0.82 15.32 28 Brain Age 2: More Training in Minutes a Day DS 2005 Puzzle Nintendo 3.44 5.36 5.32 1.18 15.3 29 Gran Turismo 3: A-Spec PS2 2001 Racing Sony Computer Entertainment 6.85 5.09 1.87 1.16 14.98 30 Call of Duty: Modern Warfare 3 X360 2011 Shooter Activision 9.03 4.28 0.13 1.32 14.76 31 Pokémon Yellow: Special Pikachu Edition GB 1998 Role-Playing Nintendo 5.89 5.04 3.12 0.59 14.64 32 Call of Duty: Black Ops X360 2010 Shooter Activision 9.67 3.73 0.11 1.13 14.64 33 Pokemon X/Pokemon Y 3DS 2013 Role-Playing Nintendo 5.17 4.05 4.34 0.79 14.35 34 Call of Duty: Black Ops 3 PS4 2015 Shooter Activision 5.77 5.81 0.35 2.31 14.24 35 Call of Duty: Black Ops II PS3 2012 Shooter Activision 4.99 5.88 0.65 2.52 14.03 36 Call of Duty: Black Ops II X360 2012 Shooter Activision 8.25 4.3 0.07 1.12 13.73 37 Call of Duty: Modern Warfare 2 X360 2009 Shooter Activision 8.52 3.63 0.08 1.29 13.51 38 Call of Duty: Modern Warfare 3 PS3 2011 Shooter Activision 5.54 5.82 0.49 1.62 13.46 39 Grand Theft Auto III PS2 2001 Action Take-Two Interactive 6.99 4.51 0.3 1.3 13.1 40 Super Smash Bros. Brawl Wii 2008 Fighting Nintendo 6.75 2.61 2.66 1.02 13.04 41 Call of Duty: Black Ops PS3 2010 Shooter Activision 5.98 4.44 0.48 1.83 12.73 42 Animal Crossing: Wild World DS 2005 Simulation Nintendo 2.55 3.52 5.33 0.88 12.27 43 Mario Kart 7 3DS 2011 Racing Nintendo 4.74 3.91 2.67 0.89 12.21 44 Halo 3 X360 2007 Shooter Microsoft Game Studios 7.97 2.83 0.13 1.21 12.14 45 Grand Theft Auto V PS4 2014 Action Take-Two Interactive 3.8 5.81 0.36 2.02 11.98 46 Pokemon HeartGold/Pokemon SoulSilver DS 2009 Action Nintendo 4.4 2.77 3.96 0.77 11.9 47 Super Mario 64 N64 1996 Platform Nintendo 6.91 2.85 1.91 0.23 11.89 48 Gran Turismo 4 PS2 2004 Racing Sony Computer Entertainment 3.01 0.01 1.1 7.53 11.66 49 Super Mario Galaxy Wii 2007 Platform Nintendo 6.16 3.4 1

In [0]:
df = spark.read.format("csv").option("header", "true").load("dbfs:/FileStore/shared_uploads/joaquin.bembhy@mip.polimi.it/vgsales.csv")

In [0]:
#lets see the dataframe
display(df)

Rank Name Platform Year Genre Publisher NA_Sales EU_Sales JP_Sales Other_Sales Global_Sales 1 Wii Sports Wii 2006 Sports Nintendo 41.49 29.02 3.77 8.46 82.74 2 Super Mario Bros. NES 1985 Platform Nintendo 29.08 3.58 6.81 0.77 40.24 3 Mario Kart Wii Wii 2008 Racing Nintendo 15.85 12.88 3.79 3.31 35.82 4 Wii Sports Resort Wii 2009 Sports Nintendo 15.75 11.01 3.28 2.96 33 5 Pokemon Red/Pokemon Blue GB 1996 Role-Playing Nintendo 11.27 8.89 10.22 1 31.37 6 Tetris GB 1989 Puzzle Nintendo 23.2 2.26 4.22 0.58 30.26 7 New Super Mario Bros. DS 2006 Platform Nintendo 11.38 9.23 6.5 2.9 30.01 8 Wii Play Wii 2006 Misc Nintendo 14.03 9.2 2.93 2.85 29.02 9 New Super Mario Bros. Wii Wii 2009 Platform Nintendo 14.59 7.06 4.7 2.26 28.62 10 Duck Hunt NES 1984 Shooter Nintendo 26.93 0.63 0.28 0.47 28.31 11 Nintendogs DS 2005 Simulation Nintendo 9.07 11 1.93 2.75 24.76 12 Mario Kart DS DS 2005 Racing Nintendo 9.81 7.57 4.13 1.92 23.42 13 Pokemon Gold/Pokemon Silver GB 1999 Role-Playing Nintendo 9 6.18 7.2 0.71 23.1 14 Wii Fit Wii 2007 Sports Nintendo 8.94 8.03 3.6 2.15 22.72 15 Wii Fit Plus Wii 2009 Sports Nintendo 9.09 8.59 2.53 1.79 22 16 Kinect Adventures! X360 2010 Misc Microsoft Game Studios 14.97 4.94 0.24 1.67 21.82 17 Grand Theft Auto V PS3 2013 Action Take-Two Interactive 7.01 9.27 0.97 4.14 21.4 18 Grand Theft Auto: San Andreas PS2 2004 Action Take-Two Interactive 9.43 0.4 0.41 10.57 20.81 19 Super Mario World SNES 1990 Platform Nintendo 12.78 3.75 3.54 0.55 20.61 20 Brain Age: Train Your Brain in Minutes a Day DS 2005 Misc Nintendo 4.75 9.26 4.16 2.05 20.22 21 Pokemon Diamond/Pokemon Pearl DS 2006 Role-Playing Nintendo 6.42 4.52 6.04 1.37 18.36 22 Super Mario Land GB 1989 Platform Nintendo 10.83 2.71 4.18 0.42 18.14 23 Super Mario Bros. 3 NES 1988 Platform Nintendo 9.54 3.44 3.84 0.46 17.28 24 Grand Theft Auto V X360 2013 Action Take-Two Interactive 9.63 5.31 0.06 1.38 16.38 25 Grand Theft Auto: Vice City PS2 2002 Action Take-Two Interactive 8.41 5.49 0.47 1.78 16.15 26 Pokemon Ruby/Pokemon Sapphire GBA 2002 Role-Playing Nintendo 6.06 3.9 5.38 0.5 15.85 27 Pokemon Black/Pokemon White DS 2010 Role-Playing Nintendo 5.57 3.28 5.65 0.82 15.32 28 Brain Age 2: More Training in Minutes a Day DS 2005 Puzzle Nintendo 3.44 5.36 5.32 1.18 15.3 29 Gran Turismo 3: A-Spec PS2 2001 Racing Sony Computer Entertainment 6.85 5.09 1.87 1.16 14.98 30 Call of Duty: Modern Warfare 3 X360 2011 Shooter Activision 9.03 4.28 0.13 1.32 14.76 31 Pokémon Yellow: Special Pikachu Edition GB 1998 Role-Playing Nintendo 5.89 5.04 3.12 0.59 14.64 32 Call of Duty: Black Ops X360 2010 Shooter Activision 9.67 3.73 0.11 1.13 14.64 33 Pokemon X/Pokemon Y 3DS 2013 Role-Playing Nintendo 5.17 4.05 4.34 0.79 14.35 34 Call of Duty: Black Ops 3 PS4 2015 Shooter Activision 5.77 5.81 0.35 2.31 14.24 35 Call of Duty: Black Ops II PS3 2012 Shooter Activision 4.99 5.88 0.65 2.52 14.03 36 Call of Duty: Black Ops II X360 2012 Shooter Activision 8.25 4.3 0.07 1.12 13.73 37 Call of Duty: Modern Warfare 2 X360 2009 Shooter Activision 8.52 3.63 0.08 1.29 13.51 38 Call of Duty: Modern Warfare 3 PS3 2011 Shooter Activision 5.54 5.82 0.49 1.62 13.46 39 Grand Theft Auto III PS2 2001 Action Take-Two Interactive 6.99 4.51 0.3 1.3 13.1 40 Super Smash Bros. Brawl Wii 2008 Fighting Nintendo 6.75 2.61 2.66 1.02 13.04 41 Call of Duty: Black Ops PS3 2010 Shooter Activision 5.98 4.44 0.48 1.83 12.73 42 Animal Crossing: Wild World DS 2005 Simulation Nintendo 2.55 3.52 5.33 0.88 12.27 43 Mario Kart 7 3DS 2011 Racing Nintendo 4.74 3.91 2.67 0.89 12.21 44 Halo 3 X360 2007 Shooter Microsoft Game Studios 7.97 2.83 0.13 1.21 12.14 45 Grand Theft Auto V PS4 2014 Action Take-Two Interactive 3.8 5.81 0.36 2.02 11.98 46 Pokemon HeartGold/Pokemon SoulSilver DS 2009 Action Nintendo 4.4 2.77 3.96 0.77 11.9 47 Super Mario 64 N64 1996 Platform Nintendo 6.91 2.85 1.91 0.23 11.89 48 Gran Turismo 4 PS2 2004 Racing Sony Computer Entertainment 3.01 0.01 1.1 7.53 11.66 49 Super Mario Galaxy Wii 2007 Platform Nintendo 6.16 3.4 1

In [0]:
#lets see it's schema
df.printSchema()

root
-- Rank: string (nullable = true)
-- Name: string (nullable = true)
-- Platform: string (nullable = true)
-- Year: string (nullable = true)
-- Genre: string (nullable = true)
-- Publisher: string (nullable = true)
-- NA_Sales: string (nullable = true)
-- EU_Sales: string (nullable = true)
-- JP_Sales: string (nullable = true)
-- Other_Sales: string (nullable = true)
-- Global_Sales: string (nullable = true)

We have many numerical columns as strings. We must convert them into double type

In [0]:
from pyspark.sql.functions import col
from pyspark.sql.types import IntegerType

df = df.withColumn('Rank', col('Rank').cast('Double')).\
    withColumn('Year', col('Year').cast('Integer')).\
    withColumn('NA_Sales', col('NA_Sales').cast('Double')).\
    withColumn('EU_Sales', col('EU_Sales').cast('Double')).\
    withColumn('JP_Sales', col('JP_Sales').cast('Double')).\
    withColumn('Other_Sales', col('Other_Sales').cast('Double')).\
    withColumn('Global_Sales', col('Global_Sales').cast('Double'))

df.printSchema()

root
-- Rank: double (nullable = true)
-- Name: string (nullable = true)
-- Platform: string (nullable = true)
-- Year: integer (nullable = true)
-- Genre: string (nullable = true)
-- Publisher: string (nullable = true)
-- NA_Sales: double (nullable = true)
-- EU_Sales: double (nullable = true)
-- JP_Sales: double (nullable = true)
-- Other_Sales: double (nullable = true)
-- Global_Sales: double (nullable = true)

In [0]:
#check the shape
print((df.count(), len(df.columns)))

(16598, 11)

We have 16,598 rows and 11 columns

Now we'll check for **null values**

In [0]:
from pyspark.sql.functions import isnan, when, count, col

df.select([count(when(isnan(c), c)).alias(c) for c in df.columns]).show()

+----+----+--------+----+-----+---------+--------+--------+--------+-----------+------------+
Rank|Name|Platform|Year|Genre|Publisher|NA_Sales|EU_Sales|JP_Sales|Other_Sales|Global_Sales|
+----+----+--------+----+-----+---------+--------+--------+--------+-----------+------------+
 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0|
+----+----+--------+----+-----+---------+--------+--------+--------+-----------+------------+

No null values. Let's continue.

In [0]:
df = df.dropna()
print((df.count(), len(df.columns)))

(16327, 11)

Now we have all the columns in the type that we need, and the data doesn't contain any null values. We have a clean dataset.

--- 

### **Time to work!**

In [0]:
df.cache()
display(df.describe())

summary Rank Name Platform Year Genre Publisher NA_Sales EU_Sales JP_Sales Other_Sales Global_Sales count 16327 16327 16327 16327 16327 16327 16327 16327 16327 16327 16327 mean 8292.868193789429 1942.0 2600.0 2006.4064433147546 null null 0.2654149568200114 0.14755435781224593 0.07866111349298115 0.04832547314264007 0.5402315183438995 stddev 4792.66977762406 null 0.0 5.828981114713253 null null 0.821590850362404 0.5087656986961239 0.3115569815994324 0.18988542720451174 1.5657318894622656 min 1.0 '98 Koshien 2600 1980 Action 10TACLE Studios 0.0 0.0 0.0 0.0 0.01 max 16600.0 ¡Shin Chan Flipa en colores! XOne 2020 Strategy responDESIGN 41.49 29.02 10.22 10.57 82.74

---
**Attention:** 'Year' variable represents the year of release of the game, so it doesn't represent the sales in that year, but the total sales corresponding to the year of release of the game

### 1. Calculate the 5 publishers with the highest historic 'Global_Sales'

In [0]:
global_sales_per_publisher =  df.groupBy('Publisher').sum('Global_Sales').sort(desc('sum(Global_Sales)')).withColumnRenamed('sum(Global_Sales)', 'Historic Global Sales')
display(global_sales_per_publisher.limit(5))

Publisher Historic Global Sales Nintendo 1784.4299999999982 Electronic Arts 1093.389999999996 Activision 721.4099999999981 Sony Computer Entertainment 607.2799999999988 Ubisoft 473.53999999999934

### 2. Find how many games each platform released each year (between 2000 and 2012)

In [0]:
display((df.groupBy("Platform", "Year").count()).where(col('count') >= 1).where(col("Year").between(2000,2012)).sort(asc('Year')))

Platform Year count PC 2000 7 N64 2000 60 PS2 2000 82 DC 2000 20 XB 2000 1 WS 2000 2 GBA 2000 1 PS 2000 159 GB 2000 17 PS2 2001 185 N64 2001 9 WS 2001 2 PC 2001 15 GBA 2001 107 PS 2001 91 GC 2001 22 XB 2001 33 GB 2001 10 DC 2001 8 XB 2002 158 PS 2002 20 GC 2002 152 GBA 2002 198 N64 2002 1 PC 2002 19 DC 2002 1 PS2 2002 280 PC 2003 33 GC 2003 144 GBA 2003 150 PS2 2003 256 PS 2003 3 XB 2003 189 PC 2004 31 GBA 2004 175 PSP 2004 15 PS2 2004 259 XB 2004 177 GC 2004 83 DS 2004 23 PC 2005 37 GC 2005 97 GBA 2005 135 X360 2005 18 DS 2005 118 PSP 2005 97 PS2 2005 260 XB 2005 179 X360 2006 93 Wii 2006 44 GBA 2006 39 PC 2006 52 PS3 2006 27 PS2 2006 259 XB 2006 62 DS 2006 202 GC 2006 40 PSP 2006 190 X360 2007 123 GC 2007 4 PS3 2007 90 Wii 2007 187 DC 2007 1 PSP 2007 134 GBA 2007 6 PC 2007 63 DS 2007 377 XB 2007 3 PS2 2007 214 Wii 2008 282 PC 2008 76 PSP 2008 100 X360 2008 146 PS3 2008 139 DS 2008 492 DC 2008 1 PS2 2008 191 XB 2008 1 X360 2009 173 Wii 2009 325 PS3 2009 162 PS2 2009 96 PSP 2009 161 PC 2009 109 DS 2009 405 Wii 2010 254 PS2 2010 38 X360 2010 182 PC 2010 90 PSP 2010 188 PS3 2010 181 DS 2010 326 X360 2011 206 PS3 2011 216 DS 2011 153 Wii 2011 144 PSP 2011 139 3DS 2011 116 PSV 2011 18 PC 2011 140 PS2 2011 7 PSV 2012 54 PS3 2012 148 3DS 2012 93 X360 2012 106 DS 2012 25 Wii 2012 32 PSP 2012 106 PC 2012 61 WiiU 2012 32

If not selected by default, please select line graph for clear visualization of the output.

Even though its not the best visual graph, Databricks visualization tools can be used to further modify the graph (zoom in, out, etc).

**Insights:**
- We can see how PS2 had a constant release of games (around 250 per year) during the 2000 first decade, but it started to decline by ends of it.
- DS became extremely popular after it's release, and many games started coming out quickly. But after the year 2008 it started crashing and by 2012 almost no more games were released. This is probably due to the introduction of other platforms as Wii, PS3, XBOX 360, and PSP (which is a direct substitute for it's portability).
- The Wii suffered the same destiny as the Nintendo DS, as many games were released quickly, but then it suddenly declined.
- There seems to be a tendency for a life cycle of platforms, but this should be confirmed with further exploration.

### 3. Calculate the contribution the top 10 videogames to the historic global revenue (%)

First we need to calculate the **total_global_revenue** for all the games. This would be the sum of the 'Global_Sales'

In [0]:
# Calculate the total historic global revenue
total_global_revenue = df.groupBy().sum('Global_Sales').withColumnRenamed('sum(Global_Sales)', 'Total Historic Revenue').collect()[0][0]
print('The total global revenue is', (total_global_revenue))

The total global revenue is 8820.360000000846

In [0]:
df = df.withColumn("Historic Global Sales", bround(lit(total_global_revenue), 2))
display(df.select('Name', 'Historic Global Sales'))

Name Historic Global Sales Wii Sports 8820.36 Super Mario Bros. 8820.36 Mario Kart Wii 8820.36 Wii Sports Resort 8820.36 Pokemon Red/Pokemon Blue 8820.36 Tetris 8820.36 New Super Mario Bros. 8820.36 Wii Play 8820.36 New Super Mario Bros. Wii 8820.36 Duck Hunt 8820.36 Nintendogs 8820.36 Mario Kart DS 8820.36 Pokemon Gold/Pokemon Silver 8820.36 Wii Fit 8820.36 Wii Fit Plus 8820.36 Kinect Adventures! 8820.36 Grand Theft Auto V 8820.36 Grand Theft Auto: San Andreas 8820.36 Super Mario World 8820.36 Brain Age: Train Your Brain in Minutes a Day 8820.36 Pokemon Diamond/Pokemon Pearl 8820.36 Super Mario Land 8820.36 Super Mario Bros. 3 8820.36 Grand Theft Auto V 8820.36 Grand Theft Auto: Vice City 8820.36 Pokemon Ruby/Pokemon Sapphire 8820.36 Pokemon Black/Pokemon White 8820.36 Brain Age 2: More Training in Minutes a Day 8820.36 Gran Turismo 3: A-Spec 8820.36 Call of Duty: Modern Warfare 3 8820.36 Pokémon Yellow: Special Pikachu Edition 8820.36 Call of Duty: Black Ops 8820.36 Pokemon X/Pokemon Y 8820.36 Call of Duty: Black Ops 3 8820.36 Call of Duty: Black Ops II 8820.36 Call of Duty: Black Ops II 8820.36 Call of Duty: Modern Warfare 2 8820.36 Call of Duty: Modern Warfare 3 8820.36 Grand Theft Auto III 8820.36 Super Smash Bros. Brawl 8820.36 Call of Duty: Black Ops 8820.36 Animal Crossing: Wild World 8820.36 Mario Kart 7 8820.36 Halo 3 8820.36 Grand Theft Auto V 8820.36 Pokemon HeartGold/Pokemon SoulSilver 8820.36 Super Mario 64 8820.36 Gran Turismo 4 8820.36 Super Mario Galaxy 8820.36 Pokemon Omega Ruby/Pokemon Alpha Sapphire 8820.36 Super Mario Land 2: 6 Golden Coins 8820.36 Grand Theft Auto IV 8820.36 Gran Turismo 8820.36 Super Mario 3D Land 8820.36 Gran Turismo 5 8820.36 Call of Duty: Modern Warfare 2 8820.36 Grand Theft Auto IV 8820.36 Super Mario All-Stars 8820.36 Pokemon FireRed/Pokemon LeafGreen 8820.36 Super Mario 64 8820.36 Just Dance 3 8820.36 Call of Duty: Ghosts 8820.36 Halo: Reach 8820.36 Mario Kart 64 8820.36 New Super Mario Bros. 2 8820.36 Halo 4 8820.36 Final Fantasy VII 8820.36 Call of Duty: Ghosts 8820.36 Just Dance 2 8820.36 Gran Turismo 2 8820.36 Call of Duty 4: Modern Warfare 8820.36 Donkey Kong Country 8820.36 Minecraft 8820.36 Animal Crossing: New Leaf 8820.36 Mario Party DS 8820.36 The Elder Scrolls V: Skyrim 8820.36 Super Mario Kart 8820.36 FIFA 16 8820.36 Wii Party 8820.36 Halo 2 8820.36 Mario Party 8 8820.36 Pokemon Black 2/Pokemon White 2 8820.36 FIFA Soccer 13 8820.36 The Sims 3 8820.36 GoldenEye 007 8820.36 Mario & Sonic at the Olympic Games 8820.36 Final Fantasy X 8820.36 Final Fantasy VIII 8820.36 Pokémon Platinum Version 8820.36 Pac-Man 8820.36 Grand Theft Auto: Liberty City Stories 8820.36 Super Mario Galaxy 2 8820.36 Star Wars Battlefront (2015) 8820.36 Call of Duty: Advanced Warfare 8820.36 The Legend of Zelda: Ocarina of Time 8820.36 Crash Bandicoot 2: Cortex Strikes Back 8820.36 Super Mario Bros. 2 8820.36 Super Smash Bros. for Wii U and 3DS 8820.36 Call of Duty: World at War 8820.36 Battlefield 3 8820.36 The Legend of Zelda: Twilight Princess 8820.36 Call of Duty: Black Ops 3 8820.36 Just Dance 8820.36 Battlefield 3 8820.36 Need for Speed Underground 8820.36 Tekken 3 8820.36 Crash Bandicoot 3: Warped 8820.36 Super Smash Bros. Melee 8820.36 Mario Kart 8 8820.36 Fallout 4 8820.36 Mario Kart: Double Dash!! 8820.36 Just Dance 4 8820.36 FIFA 14 8820.36 Need for Speed Underground 2 8820.36 Medal of Honor: Frontline 8820.36 Uncharted 3: Drake's Deception 8820.36 Crash Bandicoot 8820.36 Zumba Fitness 8820.36 Gears of War 2 8820.36 Uncharted 2: Among Thieves 8820.36 Call of Duty 4: Modern Warfare 8820.36 FIFA 12 8820.36 Big Brain Academy 8820.36 Red Dead Redemption 8820.36 FIFA 15 8820.36 Donkey Kong Country Returns 8820.36 The Elder Scrolls V: Skyrim 8820.36 The Legend of Zelda 8820.36 Assassin's Creed III 8820.36 Halo: Combat Evolved 8820.36 Pokémon Emerald Version 8820.36 Kingdom Hearts 8820.36 Pokémon Crystal Version 8820.36 Halo 3: ODST 8820.36 Red Dead Redemption 8820.36 Super Mario Sunshine 8820.3

Now we add the column to the DF and then we just calculate the percentage of global sales based on the two columns: **'Global_Sales'/'Historic Global Sales'**

In [0]:
df = df.withColumn("Percentage Global Sales",bround((col("Global_Sales")/col('Historic Global Sales'))*100, 3))
display(df.select('Name', 'Percentage Global Sales').limit(10))

Name Percentage Global Sales Wii Sports 0.938 Super Mario Bros. 0.456 Mario Kart Wii 0.406 Wii Sports Resort 0.374 Pokemon Red/Pokemon Blue 0.356 Tetris 0.343 New Super Mario Bros. 0.34 Wii Play 0.329 New Super Mario Bros. Wii 0.324 Duck Hunt 0.321

This percentage of the pie chart shows us of the percentage taking into account **ONLY** the top 10 games. The table will show us the contribution to the historic global revenue.

For example:
- **Pie chart**: Wii Sports represents the 22% of the total global revenues for the top 10 games
- **Table**: Wii Sports has signified almost the 1% of all the revenue from videogames in history.

### 4. In terms of historic global revenue, which platform has the highest market share (percentage %)?

In [0]:
global_sales_per_publisher = global_sales_per_publisher.withColumn("Total Historic Global Sales", bround(lit(total_global_revenue), 2))
global_sales_per_publisher = global_sales_per_publisher.withColumn("Historic Market Share", bround((col("Historic Global Sales")/col('Total Historic Global Sales'))*100, 1)).limit(10)
global_sales_per_publisher = global_sales_per_publisher.drop('Total Historic Global Sales')
display(global_sales_per_publisher)

Publisher Historic Global Sales Historic Market Share Nintendo 1784.4299999999982 20.2 Electronic Arts 1093.389999999996 12.4 Activision 721.4099999999981 8.2 Sony Computer Entertainment 607.2799999999988 6.9 Ubisoft 473.53999999999934 5.4 Take-Two Interactive 399.2999999999996 4.5 THQ 340.43999999999943 3.9 Konami Digital Entertainment 278.559999999998 3.2 Sega 270.69999999999925 3.1 Namco Bandai Games 253.65000000000083 2.9

**Ranking:**
- We can clearly see how **Nintento** has 20% of the historic total market share
- **Electronic Arts (EA)** ranks second with 12.4%
- **Activision** runs 3rd with 8.2%
- And so on...

### 5. Find the distribution of sales per year of release for all different regions (including Global Sales)

In [0]:
sales_per_year_region = df.groupBy('Year').sum('Global_Sales', 'NA_Sales', 'EU_Sales', 'JP_Sales', 'Other_Sales').withColumnRenamed('sum(Global_Sales)', 'Global Sales').withColumnRenamed('sum(NA_Sales)', 'NA Sales').withColumnRenamed('sum(EU_Sales)', 'EU Sales').withColumnRenamed('sum(JP_Sales)', 'JP Sales').withColumnRenamed('sum(Other_Sales)', 'Other Sales').sort(asc('Year'))

display(sales_per_year_region)

Year Global Sales NA Sales EU Sales JP Sales Other Sales 1980 11.379999999999999 10.590000000000003 0.6700000000000002 0.0 0.11999999999999998 1981 35.77000000000001 33.4 1.9600000000000006 0.0 0.3200000000000001 1982 28.859999999999996 26.920000000000005 1.6500000000000008 0.0 0.31000000000000016 1983 16.790000000000003 7.76 0.8000000000000003 8.1 0.13999999999999999 1984 50.360000000000014 33.28 2.0999999999999996 14.269999999999998 0.7000000000000002 1985 53.940000000000005 33.73 4.74 14.56 0.9200000000000002 1986 37.07 12.5 2.8400000000000007 19.81 1.9300000000000002 1987 21.739999999999995 8.460000000000003 1.4100000000000001 11.63 0.20000000000000004 1988 47.22 23.869999999999997 6.590000000000001 15.759999999999998 0.9900000000000001 1989 73.45 45.15 8.44 18.360000000000003 1.5000000000000002 1990 49.38999999999999 25.46 7.629999999999998 14.880000000000003 1.4000000000000004 1991 32.230000000000004 12.76 3.9499999999999993 14.780000000000001 0.7400000000000001 1992 76.15999999999998 33.87 11.710000000000003 28.91 1.6500000000000004 1993 45.98 15.120000000000001 4.6499999999999995 25.33000000000001 0.8900000000000001 1994 79.17000000000003 28.150000000000002 14.879999999999997 33.990000000000016 2.199999999999999 1995 88.10999999999991 24.82000000000001 14.89999999999998 45.750000000000014 2.6399999999999926 1996 199.14999999999995 86.75999999999999 47.259999999999984 57.43999999999997 7.689999999999979 1997 200.98000000000013 94.75000000000007 48.319999999999986 48.86999999999997 9.129999999999978 1998 256.46999999999963 128.35999999999999 66.90000000000012 50.04 11.02999999999994 1999 251.27000000000018 126.06000000000004 62.67000000000003 52.34 10.049999999999951 2000 201.5600000000002 94.49000000000005 52.75000000000003 42.770000000000046 11.619999999999937 2001 331.4699999999991 173.9800000000004 94.88999999999986 39.85999999999999 22.760000000000186 2002 395.5199999999983 216.1900000000001 109.74000000000032 41.76000000000002 27.280000000000253 2003 357.8499999999989 193.59000000000066 103.8100000000003 34.20000000000003 26.010000000000247 2004 419.30999999999864 222.5900000000004 107.32000000000035 41.64999999999999 47.28999999999983 2005 459.9399999999976 242.6100000000005 121.94000000000041 54.28000000000001 40.57999999999983 2006 521.0399999999917 263.11999999999887 129.23999999999984 73.72999999999998 54.429999999999836 2007 611.1299999999934 312.04999999999836 160.49999999999972 60.290000000000106 77.60000000000102 2008 678.8999999999952 351.43999999999915 184.3999999999998 60.26000000000002 82.39000000000141 2009 667.2999999999947 338.8499999999989 191.5899999999998 61.88999999999998 74.77000000000145 2010 600.4499999999948 304.24 176.73000000000016 59.490000000000215 59.89999999999995 2011 515.9899999999969 241.0600000000009 167.44000000000023 53.040000000000084 54.38999999999983 2012 363.5399999999984 154.96000000000004 118.78000000000002 51.74000000000012 37.82 2013 368.10999999999865 154.7700000000001 125.80000000000004 47.59000000000005 39.82000000000001 2014 337.0499999999985 131.9700000000002 125.6500000000001 39.460000000000086 40.02 2015 264.43999999999795 102.81999999999992 97.71000000000002 33.72000000000013 30.010000000000133 2016 70.93000000000013 22.66000000000006 26.760000000000066 13.699999999999969 7.749999999999982 2017 0.05 0.0 0.0 0.05 0.0 2020 0.29 0.27 0.0 0.0 0.02

**Insights:**
- NA accounts for a huge amount of total sales, specially with their games released after the year 2000.
- JP sales have stayed relatively constant throughout the last decade, compared to other regions
- Sales released around the year 2008 are the ones which have given the highest revenue, but then it declined for all of the regions

### 6. Compute the sales difference between the NA_Sales and the other regions for each year

In [0]:
def difference(x,y):
    return x-y

from pyspark.sql.functions import udf
dif = udf(difference, DoubleType())

In [0]:
df_dif = sales_per_year_region.select('Year', dif(sales_per_year_region['NA Sales'], sales_per_year_region['EU Sales']), dif(sales_per_year_region['NA Sales'], sales_per_year_region['JP Sales']), dif(sales_per_year_region['NA Sales'], sales_per_year_region['Other Sales'])).withColumnRenamed('difference(NA Sales, EU Sales)', 'NA - EU').withColumnRenamed('difference(NA Sales, JP Sales)', 'NA - JP').withColumnRenamed('difference(NA Sales, Other Sales)', 'NA - Other')

display(df_dif)

Year NA - EU NA - JP NA - Other 1980 9.920000000000003 10.590000000000003 10.470000000000004 1981 31.439999999999998 33.4 33.08 1982 25.270000000000003 26.920000000000005 26.610000000000007 1983 6.959999999999999 -0.33999999999999986 7.62 1984 31.18 19.010000000000005 32.58 1985 28.989999999999995 19.169999999999995 32.809999999999995 1986 9.66 -7.309999999999999 10.57 1987 7.0500000000000025 -3.169999999999998 8.260000000000003 1988 17.279999999999998 8.11 22.88 1989 36.71 26.789999999999996 43.65 1990 17.830000000000002 10.579999999999998 24.060000000000002 1991 8.81 -2.0200000000000014 12.02 1992 22.159999999999997 4.959999999999997 32.22 1993 10.470000000000002 -10.210000000000008 14.23 1994 13.270000000000005 -5.840000000000014 25.950000000000003 1995 9.92000000000003 -20.930000000000003 22.180000000000017 1996 39.50000000000001 29.32000000000002 79.07000000000001 1997 46.430000000000085 45.8800000000001 85.62000000000009 1998 61.459999999999866 78.32 117.33000000000004 1999 63.390000000000015 73.72000000000004 116.01000000000009 2000 41.74000000000002 51.720000000000006 82.87000000000012 2001 79.09000000000053 134.1200000000004 151.2200000000002 2002 106.44999999999979 174.4300000000001 188.90999999999985 2003 89.78000000000036 159.3900000000006 167.5800000000004 2004 115.27000000000005 180.9400000000004 175.30000000000058 2005 120.67000000000009 188.3300000000005 202.03000000000065 2006 133.87999999999903 189.3899999999989 208.68999999999903 2007 151.54999999999865 251.75999999999826 234.44999999999735 2008 167.03999999999934 291.17999999999915 269.04999999999774 2009 147.25999999999908 276.9599999999989 264.0799999999974 2010 127.50999999999985 244.7499999999998 244.34000000000006 2011 73.62000000000069 188.02000000000083 186.6700000000011 2012 36.18000000000002 103.21999999999991 117.14000000000004 2013 28.970000000000056 107.18000000000004 114.95000000000009 2014 6.320000000000093 92.5100000000001 91.95000000000019 2015 5.1099999999999 69.0999999999998 72.80999999999979 2016 -4.100000000000005 8.960000000000091 14.910000000000078 2017 0.0 -0.05 0.0 2020 0.27 0.27 0.25

**Insights:**
- Before the year 1995, Japan had games with higher sales than North America, but then NA games sales boomed.
- Difference in sales between NA and other regions before 1995 weren't that high, but later the gap increased hugely.
- After the year 2000, NA difference with JP and Other parts of the world was much higher than the difference between EU.
- The gap started decreasing after 2008 (but looking at previous graphs, general sales have to, so this doesnt mean much)

### 7. In how many regions did each game sales led?

Let's retrieve a column with the number of sales for the regions that the game most sold in

In [0]:
maxReg = df.select('Name', 'NA_Sales', 'EU_Sales', 'JP_Sales', 'Other_Sales',  greatest(col("NA_Sales"), col("EU_Sales"), col("JP_Sales"), col("Other_Sales")).alias("max"))
display(maxReg)

Name NA_Sales EU_Sales JP_Sales Other_Sales max Wii Sports 41.49 29.02 3.77 8.46 41.49 Super Mario Bros. 29.08 3.58 6.81 0.77 29.08 Mario Kart Wii 15.85 12.88 3.79 3.31 15.85 Wii Sports Resort 15.75 11.01 3.28 2.96 15.75 Pokemon Red/Pokemon Blue 11.27 8.89 10.22 1.0 11.27 Tetris 23.2 2.26 4.22 0.58 23.2 New Super Mario Bros. 11.38 9.23 6.5 2.9 11.38 Wii Play 14.03 9.2 2.93 2.85 14.03 New Super Mario Bros. Wii 14.59 7.06 4.7 2.26 14.59 Duck Hunt 26.93 0.63 0.28 0.47 26.93 Nintendogs 9.07 11.0 1.93 2.75 11.0 Mario Kart DS 9.81 7.57 4.13 1.92 9.81 Pokemon Gold/Pokemon Silver 9.0 6.18 7.2 0.71 9.0 Wii Fit 8.94 8.03 3.6 2.15 8.94 Wii Fit Plus 9.09 8.59 2.53 1.79 9.09 Kinect Adventures! 14.97 4.94 0.24 1.67 14.97 Grand Theft Auto V 7.01 9.27 0.97 4.14 9.27 Grand Theft Auto: San Andreas 9.43 0.4 0.41 10.57 10.57 Super Mario World 12.78 3.75 3.54 0.55 12.78 Brain Age: Train Your Brain in Minutes a Day 4.75 9.26 4.16 2.05 9.26 Pokemon Diamond/Pokemon Pearl 6.42 4.52 6.04 1.37 6.42 Super Mario Land 10.83 2.71 4.18 0.42 10.83 Super Mario Bros. 3 9.54 3.44 3.84 0.46 9.54 Grand Theft Auto V 9.63 5.31 0.06 1.38 9.63 Grand Theft Auto: Vice City 8.41 5.49 0.47 1.78 8.41 Pokemon Ruby/Pokemon Sapphire 6.06 3.9 5.38 0.5 6.06 Pokemon Black/Pokemon White 5.57 3.28 5.65 0.82 5.65 Brain Age 2: More Training in Minutes a Day 3.44 5.36 5.32 1.18 5.36 Gran Turismo 3: A-Spec 6.85 5.09 1.87 1.16 6.85 Call of Duty: Modern Warfare 3 9.03 4.28 0.13 1.32 9.03 Pokémon Yellow: Special Pikachu Edition 5.89 5.04 3.12 0.59 5.89 Call of Duty: Black Ops 9.67 3.73 0.11 1.13 9.67 Pokemon X/Pokemon Y 5.17 4.05 4.34 0.79 5.17 Call of Duty: Black Ops 3 5.77 5.81 0.35 2.31 5.81 Call of Duty: Black Ops II 4.99 5.88 0.65 2.52 5.88 Call of Duty: Black Ops II 8.25 4.3 0.07 1.12 8.25 Call of Duty: Modern Warfare 2 8.52 3.63 0.08 1.29 8.52 Call of Duty: Modern Warfare 3 5.54 5.82 0.49 1.62 5.82 Grand Theft Auto III 6.99 4.51 0.3 1.3 6.99 Super Smash Bros. Brawl 6.75 2.61 2.66 1.02 6.75 Call of Duty: Black Ops 5.98 4.44 0.48 1.83 5.98 Animal Crossing: Wild World 2.55 3.52 5.33 0.88 5.33 Mario Kart 7 4.74 3.91 2.67 0.89 4.74 Halo 3 7.97 2.83 0.13 1.21 7.97 Grand Theft Auto V 3.8 5.81 0.36 2.02 5.81 Pokemon HeartGold/Pokemon SoulSilver 4.4 2.77 3.96 0.77 4.4 Super Mario 64 6.91 2.85 1.91 0.23 6.91 Gran Turismo 4 3.01 0.01 1.1 7.53 7.53 Super Mario Galaxy 6.16 3.4 1.2 0.76 6.16 Pokemon Omega Ruby/Pokemon Alpha Sapphire 4.23 3.37 3.08 0.65 4.23 Super Mario Land 2: 6 Golden Coins 6.16 2.04 2.69 0.29 6.16 Grand Theft Auto IV 6.76 3.1 0.14 1.03 6.76 Gran Turismo 4.02 3.87 2.54 0.52 4.02 Super Mario 3D Land 4.89 2.99 2.13 0.78 4.89 Gran Turismo 5 2.96 4.88 0.81 2.12 4.88 Call of Duty: Modern Warfare 2 4.99 3.69 0.38 1.63 4.99 Grand Theft Auto IV 4.76 3.76 0.44 1.62 4.76 Super Mario All-Stars 5.99 2.15 2.12 0.29 5.99 Pokemon FireRed/Pokemon LeafGreen 4.34 2.65 3.15 0.35 4.34 Super Mario 64 5.08 3.11 1.25 0.98 5.08 Just Dance 3 6.05 3.15 0.0 1.07 6.05 Call of Duty: Ghosts 6.72 2.63 0.04 0.82 6.72 Halo: Reach 7.03 1.98 0.08 0.78 7.03 Mario Kart 64 5.55 1.94 2.23 0.15 5.55 New Super Mario Bros. 2 3.66 3.07 2.47 0.63 3.66 Halo 4 6.63 2.36 0.04 0.73 6.63 Final Fantasy VII 3.01 2.47 3.28 0.96 3.28 Call of Duty: Ghosts 4.09 3.73 0.38 1.38 4.09 Just Dance 2 5.84 2.89 0.01 0.78 5.84 Gran Turismo 2 3.88 3.42 1.69 0.5 3.88 Call of Duty 4: Modern Warfare 5.91 2.38 0.13 0.9 5.91 Donkey Kong Country 4.36 1.71 3.0 0.23 4.36 Minecraft 5.58 2.83 0.02 0.77 5.58 Animal Crossing: New Leaf 2.01 2.32 4.36 0.41 4.36 Mario Party DS 4.46 1.88 1.98 0.7 4.46 The Elder Scrolls V: Skyrim 5.03 2.86 0.1 0.85 5.03 Super Mario Kart 3.54 1.24 3.81 0.18 3.81 FIFA 16 1.11 6.06 0.06 1.26 6.06 Wii Party 1.79 3.53 2.49 0.68 3.53 Halo 2 6.82 1.53 0.05 0.08 6.82 Mario Party 8 3.81 2.3 1.58 0.73 3.81 Pokemon Black 2/Pokemon White 2 2.91 1.86 3.14 0.43 3.14 FIFA Soccer 13 1.06 5.05 0.13 2.01 5.05 The Sims 3 0.98 6.42 0.0 0.71 6.42 GoldenEye 007 5.8 2.01 0.13 0.15 5.8 Mario & Sonic at the Olympic Games 2.58 3.9 0.66 0.91 3.9 Fina

Now lets define a function that will compare each region to the max region value.

In [0]:
def maxRegion (a,b, c, d, e):
    if a == e:
        return "NA"
    elif b == e:
        return 'EU'
    elif c == e:
        return 'JP'
    else:
        return 'Other'

In [0]:
maxRegion = udf(maxRegion, StringType())

At last, apply the function to the select query and then count the values for each region. Display the distribution

In [0]:
df_maxReg = maxReg.select('*', maxRegion(maxReg['NA_Sales'], maxReg['EU_Sales'], maxReg['JP_Sales'], maxReg['Other_Sales'], maxReg['max'])).withColumnRenamed('maxRegion(NA_Sales, EU_Sales, JP_Sales, Other_Sales, max)', 'Region with max sales')
#we can drop the column max as we don't need it anymore
df_maxReg.drop('max')
max_per_region = df_maxReg.groupBy('Region with max sales').count().sort(desc('count')).withColumnRenamed('count', 'Total games')
display(max_per_region)

Region with max sales Total games NA 9925 JP 3986 EU 2340 Other 76

**Insights:**
1. North America: almost 10k games which sold the most in this region
2. JP: almost 4k games that sol the most in this region
3. EU: above 2k games that sold the most in this region
4. Other: only 76 games that sold the most

### 8. Best selling game for each region

First we must create a window partitioned by the "Regionn with max sales" column

In [0]:
from pyspark.sql import Window
w = Window.partitionBy('Region with max sales')

Now we compute the maximum of each group and then filter out the rows so that the value in the max column in df_maxReg is equal to the max of the window.

In [0]:
max_per_region = df_maxReg.withColumn('maxMax', max('max').over(w)).where(col('max') == col('maxMax')).drop('maxMax').sort(desc('max'))
top_games_per_region = max_per_region.select('Name', 'Region with max sales', 'max').withColumnRenamed('Region with max sales', 'Region').withColumnRenamed('max', 'Sales').withColumnRenamed('Name', 'Game')
display(top_games_per_region)

Game Region Sales Wii Sports NA 41.49 Nintendogs EU 11.0 Grand Theft Auto: San Andreas Other 10.57 Pokemon Black/Pokemon White JP 5.65

**Most sold games per region:**
- NA: Wii Sports
- EU: Nintendogs
- Other: Grand Theft Auto: San Andreas
- JP: Pokemon Black/Pokemon White

### 9. Perform a Join operation to visualize how much did the top performing games sold in each region.

In [0]:
df_join = df.join(top_games_per_region, df.Name == top_games_per_region.Game)
df_join2 = df_join.select('Rank', 'Name', 'Platform', 'Year', 'Region', 'NA_Sales', 'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales').withColumnRenamed('Region', 'Region with max sales').sort(asc('Rank'))

In [0]:
display(df_join2)

Rank Name Platform Year Region with max sales NA_Sales EU_Sales JP_Sales Other_Sales Global_Sales 1.0 Wii Sports Wii 2006 NA 41.49 29.02 3.77 8.46 82.74 11.0 Nintendogs DS 2005 EU 9.07 11.0 1.93 2.75 24.76 18.0 Grand Theft Auto: San Andreas PS2 2004 Other 9.43 0.4 0.41 10.57 20.81 27.0 Pokemon Black/Pokemon White DS 2010 JP 5.57 3.28 5.65 0.82 15.32 875.0 Grand Theft Auto: San Andreas XB 2005 Other 1.26 0.61 0.0 0.09 1.95 2122.0 Grand Theft Auto: San Andreas PC 2005 Other 0.0 0.92 0.0 0.05 0.98 9829.0 Grand Theft Auto: San Andreas X360 2008 Other 0.08 0.03 0.0 0.01 0.12

**Insights:**
- We can see that Grand Theft Auto has been developed for different platforms, but the region in which it sold the most was 'Other' (highest rank)
- Worlwide, Wii Sports is the highest sold game, while Nintendogs is Ranked 11th, Grand Theft Auto 18th (for PS2) and Pokemon 27th.